src/main/java/net/minestom/server/entity/EntityProjectile.java

```diff
@@ -111,26 +150,22 @@ public void tick(long time) {
             setNoGravity(true);
             onStuck();
         } else {
-            if (!super.onGround) {
-                return;
-            }
-            super.onGround = false;
-            setNoGravity(false);
-            onUnstuck();
+            onHit(((State.HitEntity) state).entity);
+            remove();
         }
     }
 
     /**
-     * Checks whether an arrow is stuck in block / hit an entity.
+     * Checks whether a projectile is stuck in block / hit an entity.
      *
      * @param pos    position right before current tick.
      * @param posNow position after current tick.
-     * @return if an arrow is stuck in block / hit an entity.
+     * @return current state of the projectile.
      */
     @SuppressWarnings("ConstantConditions")
-    private boolean isStuck(Pos pos, Pos posNow) {
+    private State getState(Pos pos, Pos posNow) {
         if (pos.samePoint(posNow)) {
-            return true;
+            return State.StuckInBlock;
```

Comment:
This shouldn't always return `State.StuckInBlock`, because projectiles can never unstuck this way. It should check if the block is still solid, and if it isn't, return `State.Flying`.